Document CustomResourceDefinition as stable (#22754)
kubernetes/website: content/en/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions.md

- roycaihw
- sttts
content_type: task
min-kubernetes-server-version: 1.16
weight: 20
---

<!-- overview -->
This page shows how to install a
[custom resource](/docs/concepts/extend-kubernetes/api-extension/custom-resources/)
into the Kubernetes API by creating a
[CustomResourceDefinition](/docs/reference/generated/kubernetes-api/{{< param "version" >}}/#customresourcedefinition-v1-apiextensions-k8s-io).

## {{% heading "prerequisites" %}}

{{< include "task-tutorial-prereqs.md" >}} {{< version-check >}}
If you are using an older version of Kubernetes that is still supported, switch to

> the documentation for that version to see advice that is relevant for your cluster.

Reviewer: Helpful! 👍

- roycaihw
- sttts
content_type: task

> min-kubernetes-server-version: 1.16

Reviewer 2: I agree. This statement implies that 1.16+ is required. Is there anything in this task that requires
v1.16 in order for the task to run?

> min-kubernetes-server-version: 1.16

Author: > Is there anything in this task that requires v1.16 in order for the task to run?

Yes: before v1.16 the CRD API was beta and the stable CRD API wasn't available.

Maybe we can tweak the version check shortcode to highlight when it makes sense to look at the documentation for a previous Kubernetes versioN?

> min-kubernetes-server-version: 1.16

Reviewer 2: > 
> 
> > Is there anything in this task that requires v1.16 in order for the task to run?
> 
> Yes: before v1.16 the CRD API was beta and the stable CRD API wasn't available.
> 

So, if you remove all references to `v1beta1` in the text then the version requirement makes more sense.
Though, perhaps it also makes sense to add a note about the switch from v1beta1 to v1 for readers that
are running a previous version of Kubernetes (taken from the original text).
Investigating changes to the version shortcode is a separate issue.

> min-kubernetes-server-version: 1.16

Reviewer 3: No. I don't agree. Please consider downstream distributions and existing deployments. I don't think all users are closely following the upstream changes. We should use the existing text "Make sure your Kubernetes cluster has a master version of 1.16.0 or higher to use apiextensions.k8s.io/v1, or 1.7.0 or higher for apiextensions.k8s.io/v1beta1". When 1.20 is out, we can safely remove the text about things before 1.16.

> min-kubernetes-server-version: 1.16

Author: @tengqm we provide copies of documentation for all supported releases.

I want to tell the reader that if they're on a previous release the can switch to a previous documentation version. If that's not clear enough, we can make the text about that bold or otherwise highlight it.

> min-kubernetes-server-version: 1.16

Author: > Extend the Kubernetes API with stable CustomResourceDefinitions

I wouldn't change the page title, but I can add an introduction that explains there used to be a beta version of the API and what that means for you.

Violations of the structural schema rules are reported in the `NonStructural` condition in the CustomResourceDefinition.

Structural schemas are required for `apiextensions.k8s.io/v1`. If you use a `apiextensions.k8s.io/v1beta1` CustomResourceDefinitions, that disables the following features:

* [Validation schema publishing](/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#publish-validation-schema-in-openapi-v2)
* [Webhook conversion](/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#webhook-conversion)
* [Pruning](#preserving-unknown-fields)

### Pruning versus preserving unknown fields {#preserving-unknown-fields}

> CustomResourceDefinitions traditionally store any (possibly validated) JSON as is in etcd. This means that unspecified fields (if there is a [OpenAPI v3.0 validation schema](/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation) at all) are persisted. This is in contrast to native Kubernetes resources such as a pod where unknown fields are dropped before being persisted to etcd. This is called _pruning_ of unknown fields.

Reviewer: Normally I stand with the I Hate Hyphens and Commas club (of which I am the treasurer), but in this case I think both improve clarity:
```suggestion
CustomResourceDefinitions traditionally store any (possibly validated) JSON as-is in etcd. This means that unspecified fields (if there is a [OpenAPI v3.0 validation schema](/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation) at all) are persisted. Contrast this behavior with default Kubernetes resources, such as a pod where unknown fields are dropped before being persisted to etcd. This is called _pruning_ of unknown fields.
```

Author: Noted

Violations of the structural schema rules are reported in the `NonStructural` condition in the CustomResourceDefinition.

> Structural schemas are required for `apiextensions.k8s.io/v1`. If you use a `apiextensions.k8s.io/v1beta1` CustomResourceDefinitions, that disables the following features:

Reviewer 4: Do we have deprecation policy about how long we will wait before removing deprecated API documentation? Since you removed the v1beta1 example in "Create a CustomResourceDefinition", I wonder if we should remove this sentence as well.

Author: There's no policy I know of, but that feedback makes sense.

Reviewer 3: The revised text is confusing. 

> that disables the following features

What is "that" referring to? The "structural schema"? or the action of using an old version, or the old CRD?

Author: I think you're reviewing superseded commits @tengqm - if I'm wrong, please clarify what should change.

For CustomResourceDefinitions created in `apiextensions.k8s.io/v1`, [structural OpenAPI v3 validation schemas](#specifying-a-structural-schema) are required and pruning is enabled and cannot be disabled.

{{< tabs name="CustomResourceDefinition_pruning" >}}
{{% tab name="apiextensions.k8s.io/v1" %}}

For CustomResourceDefinitions created in `apiextensions.k8s.io/v1`, [structural OpenAPI v3 validation schemas](#specifying-a-structural-schema) are required and pruning is enabled and cannot be disabled (note that CRDs converted from `apiextensions.k8s.io/v1beta1` to `apiextensions.k8s.io/v1` might lack structural schemas, and `spec.preserveUnknownFields` might be `true`).

{{% /tab %}}
{{% tab name="apiextensions.k8s.io/v1beta1" %}}

For CustomResourceDefinitions created in `apiextensions.k8s.io/v1beta1`, if a [structural OpenAPI v3 validation schema](#specifying-a-structural-schema) is defined (either in the global `spec.validation.openAPIV3Schema` in `apiextensions.k8s.io/v1beta1` or for each version) in a CustomResourceDefinition, pruning can be enabled by setting `spec.preserveUnknownFields` to `false`.

{{% /tab %}}
{{% /tabs %}}

> If pruning is enabled, unspecified fields in CustomResources on creation and on update are dropped.

Reviewer 4: Let's keep this line, since it explains what pruning means, and the following few lines demonstrate the behavior

Reviewer 2: Hello @sftim @roycaihw . 
Are there more comments to address or have all the changes been made?

Reviewer 3: This feature is something we want to highlight rather than drop out. When developing operators, a common mistake is that we specify some fields not known to the controller or vice versa.

kubectl create --validate=false -f my-crontab.yaml -o yaml
```

> you output is similar to:

Reviewer: ```suggestion
your output is similar to:
```

kubectl create --validate=false -f my-crontab.yaml -o yaml
```

> you output is similar to:

Reviewer 2: nit: This section may read better as a list instead of breaking up the commands and output.
```
1. Save the following YAML to my-crontab.yaml
2. Create the CronTab resource
3. Output
```

nit: `your output is similar to`

  image: my-awesome-cron-image
```

Notice that the field `someRandomField` was pruned.

This example turned off client-side validation to demonstrate the API server's behavior, by adding the `--validate=false` command line option.
Because the [OpenAPI validation schemas are also published](/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#publish-validation-schema-in-openapi-v2)

> to clients, `kubectl` also checks for unknown fields and reject those objects well before they would be sent to the API server.

Author: ```suggestion
to clients, `kubectl` also checks for unknown fields and rejects those objects well before they would be sent to the API server.
```

The field `someRandomField` has been pruned.

Notice that the `kubectl create` call uses `--validate=false` to skip client-side validation.
Because the [OpenAPI validation schemas are also published](/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#publish-validation-schema-in-openapi-v2)

> to clients, kubectl also checks for unknown fields and reject those objects well before they would be sent to the API server.

Reviewer: nit
```suggestion
to clients, kubectl also checks for unknown fields and rejects those objects well before they would be sent to the API server.
```

Author: Sorted (needs squashing)

The field `someRandomField` has been pruned.

> Notice that the `kubectl create` call uses `--validate=false` to skip client-side validation. Because the [OpenAPI validation schemas are also published](/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#publish-validation-schema-in-openapi-v2) to kubectl, it will also check for unknown fields and reject those objects long before they are sent to the API server.

Reviewer 2: This seems confusing. Is this correct?

To skip client-side validation, use `kubectl create --validate=false`.
Besides client validation, `kubectl` uses the  [published OpenAPI validation schema](/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#publish-validation-schema-in-openapi-v2) to check for unknown fields and reject those objects before they are sent to the API server.

Author: I think it'd be best to fix that line outside of the scope of this PR (title: “Document CustomResourceDefinition as stable”)

{{% /tab %}}

> {{< /tabs >}}

Reviewer: @sftim 👋 Netlify is failing:

> 5:42:49 PM: Error: Error building site: "/opt/build/repo/content/en/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions.md:613:6": closing tag for shortcode 'tab' does not match start tag

I'm guessing it's because the opening statements for these lines got lost in the shuffle.

### Publish Validation Schema in OpenAPI v2

{{< feature-state state="stable" for_k8s_version="v1.16" >}}

{{< note >}}
OpenAPI v2 Publishing is available as beta since 1.15, and as alpha since 1.14. The

> `CustomResourcePublishOpenAPI` feature must be enabled, which is the case automatically for many clusters for beta features. Please refer to the [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) documentation for more information.

Reviewer 4: the feature gate lifecycle is different from an API deprecation period. I'm not sure what's the convention for feature gates, but I assume we should keep this to let users know a feature gate exists which could enable/disable this feature

Author: AIUI you _can't_ disable stable features; that's what makes them stable. If that's wrong then there's a different page we need to fix, I think.

Reviewer 3: This paragraph, IMHO, can be safely removed. 1) 1.15 is a version too old to carry on for docs; 2) the feature gate is already graduated, so it will be ignored.

### Additional printer columns

> The kubectl tool reliese on server-side output formatting. Your cluster decides which

Reviewer 2: Spelling: relies
Instead of `Your cluster`, substitute `The API server`?

Author: ```suggestion
The kubectl tool relies on server-side output formatting. Your cluster's API server decides which
```

### Subresources

{{< feature-state state="stable" for_k8s_version="v1.16" >}}

Custom resources support `/status` and `/scale` subresources.

You can disable this feature using the `CustomResourceSubresources` feature gate on

> the [kube-apiserver](/docs/admin/kube-apiserver):

Reviewer 4: same for feature gates

Reviewer 3: This one has graduated as well, so can be removed.
